src/V3Hasher.cpp

```diff
@@ -29,38 +31,398 @@ class HasherVisitor final : public AstNVisitor {
-    virtual void visit(AstParamTypeDType*) override {}
+    V3Hash hashNode(AstNode* nodep, bool hashDType, bool hashChildren, std::function<void()>&& f) {
+        if (m_cacheInUser4 && nodep->user4()) {
+            return V3Hash(nodep->user4());
+        } else {
+            VL_RESTORER(m_hash);
+            // Reset accumulator
+            m_hash = V3Hash(nodep->type());  // Node type
+            f();  // Node specific hash
+            if (hashDType && nodep != nodep->dtypep()) iterateNull(nodep->dtypep());  // Node dtype
+            if (hashChildren) iterateChildrenConst(nodep);  // Children
+            if (m_cacheInUser4) { nodep->user4(m_hash.value()); }
+            return m_hash;
+        }
+    }
+
+    // VISITORS
+
+    // Each visitor below contributes to the hash any node specific content
+    // that is not dependent on either of the following, as these are
+    // included by default by hashNode:
+    // - Node type (as given by AstNode::type())
+    // - Node dtype (unless !hashDType)
+    // - child nodes (unless !hashChildren)
+    //
+    // The hash must be stable, which means in particular it cannot rely on
+    // pointer values, or any other value that might differ between separate
+    // invocations of Verilator over the same design.
+    //
+    // Note there is a circularity problem where some child nodes can back
+    // to their ancestral nodes via member pointers, which can lead to an
+    // infinite traversal. To break this, nodes that are subject to such
+    // referencing and represent code which can reasonably be assumed not to
+    // be equivalent to any other code, are hashed either by name (e.g.:
+    // AstNodeModule), or by unique identifier (e.g.: AstNodeUOrStructDType).
+
+    //------------------------------------------------------------
+    // AstNode - Warns to help find missing cases
 
     virtual void visit(AstNode* nodep) override {
-        V3Hash thisHash;
-        if (!m_cacheInUser4 || !nodep->user4()) {
-            VL_RESTORER(m_lowerHash);
-            {
-                m_lowerHash = nodep->sameHash();
-                UASSERT_OBJ(!m_lowerHash.isIllegal(), nodep,
-                            "sameHash function undefined (returns 0) for node under CFunc.");
-                // For identical nodes, the type should be the same thus
-                // dtypep should be the same too
-                m_lowerHash = V3Hash(m_lowerHash,
-                                     V3Hash(V3Hash(nodep->type() << 6), V3Hash(nodep->dtypep())));
-                // Now update m_lowerHash for our children's (and next children) contributions
-                iterateChildrenConst(nodep);
-                // Store the hash value
-                if (m_cacheInUser4) { nodep->user4(m_lowerHash.fullValue()); }
-                thisHash = m_lowerHash;
+#if VL_DEBUG
+        UINFO(0, "WARNING! Hashing node as AstNode: " << nodep);
+#endif
+        m_hash = hashNode(nodep, true, true, [=]() {});
```

Comment:
Maybe add
     const bool HASH_DTYPE = true;
     const bool HASH_CHILDREN = true;

Then easier to read:

```suggestion
        m_hash = hashNode(nodep, HASH_DTYPE, HASH_CHILDREN, [=]() {});
```

(or the current "false" if not doing these).

Reply:
Sorry, I don't get what you mean, some trees cannot do this and are passed `false` otherwise they are cirrcular.

Reply:
I'm just suggesting a cosmetic change, that instead of "true" you indicate HASH_DTYPE or HASH_CHILDREN to show what the true is doing, that is either passing true to hashDType, or true to hashChildren.

e.g.

        m_hash = hashNode(nodep, HASH_DTYPE, HASH_CHILDREN, [=]() {});
        m_hash = hashNode(nodep, HASH_DTYPE, false, [=]() {});
        m_hash = hashNode(nodep, false, HASH_CHILDREN, [=]() {});
        m_hash = hashNode(nodep, false, false, [=]() {});

Reply:
(globally not just this one function - should have made that clearer)

Reply:
Done, but I can wholeheartedly recommend Clion where this just renders like this ;)
![image](https://user-images.githubusercontent.com/16000351/119137705-05508f80-ba39-11eb-9ef7-b9cd1d3d9c98.png)